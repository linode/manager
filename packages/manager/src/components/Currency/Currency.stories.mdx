import { ArgsTable, Canvas, Meta, Story } from '@storybook/addon-docs';
import Currency from './Currency';

<Meta title="Elements/Typography/Currency" component={Currency} />

export const CurrencyTemplate = (args) => <Currency {...args} />;

# Currency

<Canvas>
  <Story
    name="Currency"
    args={{
      quantity: 4.0,
      decimalPlaces: 2,
      wrapInParentheses: false,
    }}
    argTypes={{
      decimalPlaces: {
        control: {
          min: 0,
        },
      },
      wrapInParentheses: {
        description:
          'A boolean used to wrap the currency in parenthesis. This is normally done to indicate a negative amount or balance.',
      },
    }}
  >
    {CurrencyTemplate.bind({})}
  </Story>
</Canvas>

<ArgsTable story="Currency" exclude={['dataAttrs']} />
