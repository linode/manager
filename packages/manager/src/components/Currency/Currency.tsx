import * as React from 'react';

interface CurrencyFormatterProps {
  quantity: number;
  decimalPlaces?: number;
  wrapInParentheses?: boolean;
  dataAttrs?: Record<string, any>;
}

export const Currency: React.FC<CurrencyFormatterProps> = (props) => {
  const { quantity, wrapInParentheses, dataAttrs } = props;

  const formatter = new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: 'USD',
    minimumFractionDigits: props.decimalPlaces ?? 2,
  });

  const formattedQuantity = formatter.format(Math.abs(quantity));
  const isNegative = quantity < 0;

  let output;

  if (wrapInParentheses) {
    output = isNegative ? `-(${formattedQuantity})` : `(${formattedQuantity})`;
  } else {
    output = isNegative ? `-${formattedQuantity}` : formattedQuantity;
  }

  // eslint-disable-next-line
  return (
    <span className="notranslate" {...dataAttrs}>
      {output}
    </span>
  );
};

export default React.memo(Currency);
