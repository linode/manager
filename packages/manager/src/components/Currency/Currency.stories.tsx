import * as React from 'react';

import { Currency } from './Currency';

import type { Meta, StoryObj } from '@storybook/react';

type Story = StoryObj<typeof Currency>;

export const Default: Story = {
  args: {
    decimalPlaces: 2,
    quantity: 4.0,
    wrapInParentheses: false,
  },
  render: (args) => <Currency {...args} />,
};

const meta: Meta<typeof Currency> = {
  argTypes: {
    decimalPlaces: {
      control: {
        min: 0,
      },
    },
  },
  component: Currency,
  title: 'Components/Typography/Currency',
};

export default meta;
