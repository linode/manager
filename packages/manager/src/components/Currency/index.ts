import Currency from './Currency';
export default Currency;
