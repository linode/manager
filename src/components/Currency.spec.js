import React from 'react';
import { shallow } from 'enzyme';
import Currency from '~/components/Currency';

describe('components/Currency', () => {
  it('renders negative values', () => {
    const component = shallow(
      <Currency value={-999} />
    );

    expect(component.find('.currency').text()).toBe('($999.00)');
  });

  it('renders positive values', () => {
    const component = shallow(
      <Currency value={999} />
    );

    expect(component.find('.currency').text()).toBe('$999.00');
  });

  it('renders zero values', () => {
    const component = shallow(
      <Currency value={0} />
    );

    expect(component.find('.currency').text()).toBe('$0.00');
  });
});
