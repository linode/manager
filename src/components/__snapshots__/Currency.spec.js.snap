// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Currency should render without error 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Currency
    value={999.99}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "$999.99",
      "className": "currency",
    },
    "ref": null,
    "rendered": "$999.99",
    "type": "span",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "$999.99",
        "className": "currency",
      },
      "ref": null,
      "rendered": "$999.99",
      "type": "span",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
